external/java/src/main/java/ch/epfl/dedis/eventlog/EventLogInstance.java

```diff
@@ -172,6 +177,194 @@ public SearchResponse search(String topic, long from, long to) throws CothorityE
+                error = s;
+            }
+        }
+
+        List<Pair> buffer;
+        Boolean flushed = false;
+
+        BufferedEventLogReceiver(EventHandler h) {
+            super(h);
+            buffer = new ArrayList<>();
+        }
+
+        synchronized void flush(List<SkipBlock> prepend) {
+            SkipBlock current = null;
+            for (SkipBlock b : prepend) {
+                current = b;
+                super.receive(b);
+            }
+
+            for (Pair p : buffer) {
+                if (p.block == null) {
+                    super.error(p.error);
+                }
+                else if (p.block.getBackLinks().size() == 0) {
+                    super.error("no black links");
+                }
+                else if (current != null && Arrays.equals(p.block.getBackLinks().get(0).getId(), current.getHash())) {
+                    super.receive(p.block);
+                } else {
+                    // do nothing because this is a duplicate block
+                }
+            }
+            flushed = true;
+        }
+
+        @Override
+        public synchronized void receive(SkipBlock block) {
+            if (flushed) {
+                super.receive(block);
+            } else {
+                buffer.add(new Pair(block, null));
+            }
+        }
+
+        @Override
+        public synchronized void error(String s) {
+            if (flushed) {
+                super.error(s);
+            } else {
+                buffer.add(new Pair(null, s));
+            }
+        }
+    }
+
+    /**
+     * Register the handler to listen to new events.
+     * If an error is thrown, the caller should unsubscribe the tag.
+     *
+     * @param handler is the event handler, it will be called in a different thread.
+     * @return an integer tag that can be used to unregister the handler.
+     * @throws CothorityCommunicationException when the server fails or refuses to accept the subscription.
+     */
+    public int subscribeEvents(EventHandler handler) throws CothorityCommunicationException {
+        EventLogReceiver r = new EventLogReceiver(handler);
+        int tag = rnd.nextInt();
+        this.handlers.put(tag, r);
+        this.bc.subscribeSkipBlock(r);
+        return tag;
+    }
+
+    /**
+     * Register the handler to listen to new events from the given block.
+     * The function blocks while it reads old events from the given block.
+     * If an error is thrown, the caller should unsubscribe the tag.
+     *
+     * @param handler is the event handler, it will be called in a different thread.
+     * @param from is the block which marks the beginning of the subscription.
+     * @return an integer tag that can be used to unregister the handler.
+     * @throws CothorityCommunicationException
+     * @throws CothorityCryptoException
+     */
+    public int subscribeEvents(EventHandler handler, byte[] from) throws CothorityCommunicationException, CothorityCryptoException {
```

Comment:
`from` should be of type SkipblockId or is this for convenience ?

Reply:
same reason as https://github.com/dedis/cothority/pull/1906#discussion_r286958784